### PR TITLE
[FSDP][Perf] Avoid post-bwd pad

### PR DESCRIPTION
**Pull request by awgu**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#88583 [FSDP][Perf] Avoid post-bwd pad**

**Overview**
This PR adds an additional code path for when the `FlatParameter` must be padded to achieve an unsharded size divisible by the world size to enable the performant `_all_gather_base()` and `_reduce_scatter_base()`.
- In the padded case, we register a new pre-backward hook on the `FlatParameter` itself that _pre-allocates_ the padded unsharded gradient as a zero tensor to avoid doing so in the post-backward hook, which would (1) allocate unsharded memory in the post-backward stream and (2) require a D2D copy, which may incur nontrivial overhead and block the subsequent reduce-scatter.
    - The newly-computed unsharded gradient is added to the pre-allocated zero gradient. From a running time perspective, we tradeoff a GPU `memcpy` kernel with a GPU `add` kernel, so the running time may not decrease. From a memory perspective, we still have the computed gradient from autograd and an existing unsharded gradient in memory.
    - The `handle.prepare_gradient_for_backward()` call is refactored to be in this new pre-backward hook `_flat_param_pre_backward_hook()`. This is to delay the pre-allocation as much as possible. The new hook is guarded by `handle._ran_flat_param_pre_backward_hook` because otherwise the hook may run multiple times per backward. It is reset to `False` in the post-backward hook.
    - ~~This approach leverages the private API `torch._C._will_engine_execute_node(flat_param_acc_grad)` to check if a `FlatParameter` will compute a gradient in the current `backward()`. As a part of this, `FlatParameter`s that are not involved in the current `backward()` are not all-gathered.~~ See Notes section.
- We do not pre-allocate for the non-padded case because that would allocate an extra tensor compared to the existing behavior.
- For both the padded and non-padded cases, this PR changes the sharded gradient allocation to use `torch.empty()` instead of `torch.zeros()` to avoid a `fill` kernel.
- This PR renames `param` to `flat_param` in the post-backward hook for clarity.

**Example: RegNet-3B**
For RegNet-3B using an auto wrap policy that wraps `RegNetYLayer` and world size of 8, there are both `FlatParameter`s that need padding and those that do not.

With this PR (using a batch size of 48, `BACKWARD_PRE` prefetching, FP16 mixed precision, and activation checkpointing):
- The number of `memcpy` kernels per training iteration decreases from 234 to 6.
- [`use_orig_params=False`] The peak reserved memory decreases from 16.625 GB to 16.551 GB.
- [`use_orig_params=False`] The peak active memory decreases from 10.500 GB to 10.363 GB.
- [`use_orig_params=True`] The peak reserved memory decreases from 14.051 GB to 13.977 GB.
- [`use_orig_params=True`] The peak active memory decreases from 10.501 GB to 10.363 GB.

This reduction in peak active memory comes from saving the extra unsharded gradient allocation in the post-backward. (This suggests that the peak memory happens in the backward pass for this workload.)

<details>
    <summary>Example: 2nd ReLU Backward</summary>

As an example, we examine the 2nd ReLU backward. First, we look at the trace for the existing code:
Screen Shot 2022-11-07 at 9 23 31 AM
Screen Shot 2022-11-07 at 9 23 22 AM
We see above the `memcpy` kernel. Now, we look at the trace for the new code:
Screen Shot 2022-11-07 at 9 23 38 AM
Screen Shot 2022-11-07 at 9 25 39 AM
We see an `add` kernel above but no more `memcpy` kernel. I am not sure why exactly, but this somehow allows the reduce-scatter to run earlier. (The preceding `BUnaryFunctor` kernel right before the reduce-scatter somehow runs earlier. This behavior is consistent across multiple runs. The `BUnaryFunctor` should be the gradient pre-divide.)

</details>

<details>
<summary> _flat_param_pre_backward_hook() </summary>

This allocates a zero tensor for the padded unsharded gradient, and it runs after the normal pre-backward hook.
Screen Shot 2022-11-07 at 8 30 24 AM

</details>

**Notes**
Originally, I had the PR such that we only unshard in the pre-backward if `torch._C._will_engine_execute_node(flat_param_acc_grad)`. However, this breaks with reentrant activation checkpointing when applied as `FSDP(CheckpointWrapper(module))`. This is because when the module-level pre-backward hook runs, the `bool` will return `False`, and instead, the `FlatParameter` gradient is computed later by the explicit `torch.autograd.backward()` call:
https://github.com/pytorch/pytorch/blob/master/torch/utils/checkpoint.py#L157

This is not a major issue, but the consequence of this is that we may unnecessarily unshard `FlatParameter`s (e.g. for mult-modal models), where the pre-backward hook may run but gradients are not computed for the `FlatParameter`.

Differential Revision: [D41137855](https://our.internmc.facebook.com/intern/diff/D41137855)

### CODE REVIEW

**Comment by awgu** [comment]
@soulitzer Is it okay for FSDP to depend on `torch._C._will_engine_execute_node()`?

File: torch/distributed/fsdp/_runtime_utils.py [hunk 1]
```diff
@@ -1072,6 +1152,26 @@ def _clear_grads_if_needed(
             handle._clear_grads_if_needed()
 
 
+@no_type_check
+def _get_handles_executing_backward(
```
Comment (awgu) [inline torch/distributed/fsdp/_runtime_utils.py:1156]: Note to self: maybe rename this to `_get_handles_executing_in_backward()`.

**Comment by awgu** [comment]
I wonder if we can get the same memory savings by simply moving the padding logic to be outside the post-backward stream. The reason I say this is that the `memcpy` kernel trades off with an `add` kernel here, both of whose running times should be dominated by the memory access (i.e. they should take around the same time). By pre-allocating the padded gradient, we still have both an existing gradient and the newly-computed gradient by the autograd engine in memory simultaneously. The only way to truly save that memory is to somehow have the autograd engine allocate the computed gradient as padded in the first place.

I will investigate this.

**Comment by awgu** [comment]
@awgu has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D41137855).

**Comment by zhaojuanmao** [comment]
"This approach leverages the private API torch._C._will_engine_execute_node(flat_param_acc_grad) to check if a FlatParameter will compute a gradient in the current backward(). As a part of this, FlatParameters that are not involved in the current backward() are not all-gathered."

==> it may help multimodal model perf?

File: torch/distributed/fsdp/flat_param.py [hunk 2]
```diff
@@ -1007,28 +1008,42 @@ def reshard_grad(self):
...
 
+    @no_type_check
     def prepare_gradient_for_backward(self):
         """
         Prepares the gradient for the backward computation by saving and
         clearing any existing sharded gradient in ``.grad`` to enable computing
         a new unsharded gradient.
+
+        Args:
+            sync_gradients (bool): ``False`` if in ``no_sync()`` and ``True``
```
Comment (zhaojuanmao) [inline torch/distributed/fsdp/flat_param.py:1019]: nit: not used?

**Comment by zhaojuanmao** [comment]
> "This approach leverages the private API torch._C._will_engine_execute_node(flat_param_acc_grad) to check if a FlatParameter will compute a gradient in the current backward(). As a part of this, FlatParameters that are not involved in the current backward() are not all-gathered."
> 
> ==> it may help multimodal model perf?

did not see where it is implemented?

File: torch/distributed/fsdp/flat_param.py (hunk 2 above)
Comment (awgu) [reply torch/distributed/fsdp/flat_param.py:1019]: Good catch! Let me get rid of this.

File: torch/distributed/fsdp/_runtime_utils.py [hunk 3]
```diff
@@ -508,71 +543,102 @@ def _post_backward_hook(
...
+                    # since it owns the underlying storage.
+                    flat_param._padded_unsharded_grad.data = (
+                        flat_param._padded_unsharded_grad.to(
+                            state.mixed_precision.reduce_dtype
+                        )
+                    )
+                    flat_param.grad.data = flat_param._padded_unsharded_grad[
+                        : flat_param._unpadded_unsharded_size.numel()
+                    ]
+                else:
```
[the remark this answers is not in the document]
Comment (awgu) [reply torch/distributed/fsdp/_runtime_utils.py:573]: Agreed. Let me open a separate PR that adds an `if numel_to_pad > 0` before calling `F.pad()` to avoid calling it in the no-padding case.

**Comment by awgu** [comment]
> > "This approach leverages the private API torch._C._will_engine_execute_node(flat_param_acc_grad) to check if a FlatParameter will compute a gradient in the current backward(). As a part of this, FlatParameters that are not involved in the current backward() are not all-gathered."
> > ==> it may help multimodal model perf?
> 
> did not see where it is implemented?

Good catch! I changed the implementation after writing that because it broke with reentrant activation checkpointing. See the "Notes" section at the end that I added after fixing the breaking unit tests.

**Comment by awgu** [comment]
I prefer to compute the unsharded gradient with padding directly over this approach. I will close this PR for now.
